test/test/lock/RaceToLock.java

```diff
@@ -8,81 +8,65 @@
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
 import java.util.concurrent.locks.LockSupport;
-import java.util.Random;
 
 public class RaceToLock {
 
-    private final Lock sharedLock = new ReentrantLock(true);
-    private long sharedWaitTime = 0;
+    // Every SHARED_LOCK_INTERVAL iterations of the loop in runSemiShared, we try to lock a shared lock instead of an uncontended lock
+    private static final long SHARED_LOCK_INTERVAL = 10;
+    private static final long DURATION_MS = 4000;
 
-    private final Random random = new Random();
-    private final Lock[] locks = {
-            new ReentrantLock(true),
-            new ReentrantLock(true),
-            new ReentrantLock(true)
-    };
-    private volatile long randomWaitTime;
+    private static final Lock sharedLock = new ReentrantLock(true);
 
-    private volatile boolean exitRequested;
+    private static volatile boolean exitRequested = false;
 
-    private void doWork() {
+    private static void doWork() {
         LockSupport.parkNanos(1);
     }
 
-    private void runSharedCounter() {
+    private static void runShared() {
         while (!exitRequested) {
-            long start = System.nanoTime();
-            sharedLock.lock();
-            sharedWaitTime += System.nanoTime() - start;
-
             try {
+                sharedLock.lock();
                 doWork();
             } finally {
                 sharedLock.unlock();
             }
         }
     }
 
-    private void runRandomCounter() {
+    private static void runSemiShared() {
+        long count = 0;
+        Lock nonsharedLock = new ReentrantLock(true);
         while (!exitRequested) {
-            Lock lock = locks[random.nextInt(locks.length)];
+            Lock lock = count % SHARED_LOCK_INTERVAL == 0 ? sharedLock : nonsharedLock;
 
-            long start = System.nanoTime();
-            lock.lock();
-            randomWaitTime += System.nanoTime() - start;
             try {
+                lock.lock();
```

Comment:
Correct ReentrantLock pattern is to call `lock()` outside `try-finally`. Otherwise `unlock` may be called on an object that is not locked.

Reply:
Thanks, a3234fc49b7442dcca0f0049f8189a7afabe2d0a